File: zk_auth_lib/py/chaum_pedersen.py
import logging
import random

LOGGER = logging.getLogger()

class ChaumPedersenData():
    def __init__(self, data: object):
        LOGGER.debug("ChaumPedersenData init ")
        self._data = data
        self._data.A = pow(data.g, data.a, data.q) # g^a 
        self._data.B = pow(data.g, data.b, data.q) # g^b 
        self._data.C = pow(data.g, (data.a * data.b), data.q) # g^ab
        LOGGER.debug("Public agree (g,g^a, g^b and g^ab) \
                     =(%d,%d,%d,%d)", 
                     self._data.g,
                     self._data.A,
                     self._data.B,
                     self._data.C)

class ChaumPedersenProver(ChaumPedersenData):

    def __init__(self, data: object):
        super().__init__(data=data)

    def register(self, r: int) -> int:
        LOGGER.debug("ChaumPedersenProver register")        
        LOGGER.debug("Use Password (r) %d",r)
        self._data.r = r
        y1 = pow(self._data.g, self._data.r, self._data.q) # g^r 
        y2 = pow(self._data.B, self._data.r, self._data.q) # (g^b)^r
        LOGGER.debug("Prover sends y1 y2 (g^r, B^r)=(%d,%d)", y1, y2)
        return y1, y2
    
    def response(self, s: int) -> int:
        LOGGER.debug("ChaumPedersenProver response")                
        z = (self._data.r + self._data.a * s) % self._data.q # (r + a*s)
        LOGGER.debug("Prover computes z=r+as (mod q)=%d", z)
        return z
    

class ChaumPedersenVerifier(ChaumPedersenData):
    def __init__(self, data: object):
        super().__init__(data=data)
        # for verification
        if hasattr(data, 'r'):
            self._data.y1 = pow(self._data.g, data.r, self._data.q)
            self._data.y2 = pow(self._data.B, data.r, self._data.q)

    def challenge(self) -> int:  
        LOGGER.debug("ChaumPedersenVerifier challenge")      
        self._data.s = random.randint(1, 1000)
        LOGGER.debug("Verifier sends a challenge (s)=%d", self._data.s)
        return self._data.s

    def verify(self, z: int) -> bool:
        LOGGER.debug("ChaumPedersenVerifier verify")        
        LOGGER.debug("Verifier makes two checks using the answer (z)=%d against the challenge s", z)

        a1 = pow(self._data.g, z, self._data.q)
        LOGGER.debug("Verifier calculates a1 g^z mod q=%d", a1)

        a2 = (self._data.A**self._data.s * self._data.y1) % self._data.q
        LOGGER.debug("Verifier calculates a2: A^s y1=%d",a2)
        if a1 != a2:
            return False

        LOGGER.debug("Verifier now checks these are the same")
        LOGGER.debug("Verifier calculates B^z=%d", pow(self._data.B, z, self._data.q))
        b1 = pow(self._data.B, z, self._data.q)

        b2 = (self._data.C**self._data.s * self._data.y2) % self._data.q        
        LOGGER.debug("Verifier calculates C^s y2=%d", b2)
        if b1 != b2:
            return False
        
        return True
    

        

File: zk_auth_lib/go/chaum_pedersen.go
package chaum_pedersen

import (
	"crypto/rand"
	"fmt"
	"math/big"

	log "github.com/sirupsen/logrus"
)

type ChaumPedersenProver interface {
	Register(r *big.Int) (*big.Int, *big.Int)
	Response(s *big.Int) *big.Int
}

type ChaumPedersenVerifier interface {
	Challenge() *big.Int
	Verify(y1, y2, z *big.Int) bool
}

type chaumPedersen struct {
	g *big.Int
	q *big.Int
	a *big.Int
	b *big.Int
	A *big.Int
	B *big.Int
	C *big.Int
	r *big.Int
	s *big.Int
}

func NewChaumPedersen(g, q, a, b *big.Int) *chaumPedersen {

	A := new(big.Int).Exp(g, a, q)                      // g^a mod q
	B := new(big.Int).Exp(g, b, q)                      // g^b mod q
	C := new(big.Int).Exp(g, new(big.Int).Mul(a, b), q) // g^(a*b) mod q

	log.Trace(fmt.Printf("Public agree (g, g^a, g^b and g^ab) : %v, %v, %v, %v \n", g, A, B, C))
	return &chaumPedersen{
		g: g,
		q: q,
		a: a,
		b: b,
		A: A,
		B: B,
		C: C,
	}
}

func (cpd *chaumPedersen) Register(r *big.Int) (*big.Int, *big.Int) {
	cpd.r = r
	y1 := new(big.Int).Exp(cpd.g, cpd.r, cpd.q) // g^r mod q
	y2 := new(big.Int).Exp(cpd.B, cpd.r, cpd.q) // (g^b)^r mod q
	log.Trace(fmt.Printf("Prover computes and sends y1 y2 to Verifier (g^r, B^r)=(%v,%v)\n", y1, y2))
	return y1, y2
}

func (cpd *chaumPedersen) Response(s *big.Int) *big.Int {
	// // (r + a*s) mod q
	z := new(big.Int).Mod(cpd.r.Add(cpd.r, cpd.a.Mul(cpd.a, s)), cpd.q)
	log.Trace(fmt.Printf("Prover computes z=r+as (mod q)=%v\n", z))
	return z
}

func (cpd *chaumPedersen) Challenge() *big.Int {
	cpd.s, _ = rand.Int(rand.Reader, big.NewInt(1000))
	log.Trace(fmt.Printf("Verifier sends a challenge (s)=%v\n", cpd.s))
	return cpd.s
}

func (cpd *chaumPedersen) Verify(y1, y2, z *big.Int) bool {
	log.Trace(fmt.Printf("y1 %v ", y1))  // g^r mod q
	log.Trace(fmt.Printf("y2 %v\n", y2)) // (g^b)^r mod q
	log.Trace(fmt.Printf("Verifier makes two checks using the answer (z)=%v against the challenge (s)=%v\n", z, cpd.s))

	a1 := new(big.Int).Exp(cpd.g, z, cpd.q)
	log.Trace(fmt.Printf("Verifier calculates a1 g^z mod q=%d\n", a1))

	// A^s y1 mod q
	a2 := new(big.Int).Mod(new(big.Int).Mul(new(big.Int).Exp(cpd.A, cpd.s, nil), y1), cpd.q)
	log.Trace(fmt.Printf("Verifier calculates a2: A^s y1=%d\n", a2))

	b1 := new(big.Int).Exp(cpd.B, z, cpd.q)
	log.Trace(fmt.Printf("Verifier calculates b1: B^z=%d\n", b1))

	// C^s y2 mod q
	b2 := new(big.Int).Mod(new(big.Int).Mul(new(big.Int).Exp(cpd.C, cpd.s, nil), y2), cpd.q)
	log.Trace(fmt.Printf("Verifier calculates b2: C^s y2=%d\n", b2))

	return a1.Cmp(a2) == 0 && b1.Cmp(b2) == 0
}
